ZprimeMuMuPatMiniAodNewMC.C

```diff
@@ -798,6 +808,61 @@ void ZprimeMuMuPatMiniAodNewMC::Loop(bool debug)
 //       Part for Gen & Reco Matching                -
 //                                                   -
 //----------------------------------------------------
+//
+//
+//
+
+float ZprimeMuMuPatMiniAodNewMC::GetScaleBias(float eta, float phi, float pt, float charge)
+{
+  double shift = m_muon_scale_ratio_hist->GetBinContent(
+          m_muon_scale_ratio_hist->FindBin(eta, phi));
+  double uncertainty = m_muon_scale_ratio_hist->GetBinError(
+          m_muon_scale_ratio_hist->FindBin(eta, phi));
+ 
+
+  //rand = new TRandom3();
+  // Central value correction + gaussian uncertainty
+  if (std::abs(eta) < 1.2) {
+      shift = 0.0;
+      uncertainty = 0.025;
+  }
+  double ratio   = 1. + (rand->Gaus(shift, uncertainty) * pt / 1000./ charge);
+  return 1./ratio;
+ 
+}
+
+float ZprimeMuMuPatMiniAodNewMC::GetResultionUncert(float pt,float eta)
+{
+ 
+     //rand = new TRandom3();
+     double smearing = 0.0;
+     if (pt < 200.) {
+       smearing = 0.003;
+     } else if (pt < 500.) {
+       smearing = 0.005;
+     } else {
+       smearing = 0.01;
+     }
+     // Double smearing for muons in the endcaps
+     if (std::abs(eta) > 1.2) smearing *= 2;
+     // Return Gaussian smearing
+     double ratio = rand->Gaus(1, smearing);
+     return ratio;
+
+ 
+}
+
+TLorentzVector ZprimeMuMuPatMiniAodNewMC::GetShiftedMuon(float px, float py, float pz, float E, float ratio){
```

Comment:
Does this method leak (`new` not accompanied by `delete`)?
Could clean up with a `shared_ptr`

Reply:
Hmm, yeah, could be leaky. I'm not sure how these shared_ptr work, but if can point me to an example I can improve this.

Reply:
I'd try something a la (I always have to trial a few things to get the syntax correct for the specific constructor):
```c++
std::shared_ptr<TLorentzVector> result = std::make_shared<TLorentzVector>()
result->SetPxPyPzE(ratio*px,
                                  ratio*py,
                                  ratio*pz,
                                  ratio*E);
```
or maybe
```c++
std::shared_ptr<TLorentzVector> result = std::make_shared<TLorentzVector>(ratio*px,ratio*py,ratio*pz,ratio*E);)
```